pkg/cloudprovider/cloudprovider.go

```diff
@@ -236,6 +219,37 @@ func (c *CloudProvider) IsMachineDrifted(ctx context.Context, machine *v1alpha5.
 	return amiDrifted, nil
 }
 
+// Hydrate updates an existing instance by making sure that the machine associated with the instance
+// has the corresponding tag value for that machine
+func (c *CloudProvider) Hydrate(ctx context.Context, machine *v1alpha5.Machine) error {
+	instanceID, err := utils.ParseInstanceID(machine.Status.ProviderID)
+	if err != nil {
+		return fmt.Errorf("parsing instance id, %w", err)
+	}
+	instance, err := c.instanceProvider.GetByID(ctx, instanceID)
+	if err != nil {
+		return fmt.Errorf("getting instance for instanceID '%s', %w", instanceID, err)
+	}
+	// Update the machine with the name stored in the tag value of the instance if it exists
+	// We do this so that when we create the Machine after hydration, the machine-controller is aware
+	// that the instance already exists, so it doesn't create another one
+	if tag, ok := lo.Find(instance.Tags, func(tag *ec2.Tag) bool {
+		return aws.StringValue(tag.Key) == v1alpha5.MachineNameLabelKey
```

Comment:
brevity
```suggestion
		return lo.FromPtr(tag.Key) == v1alpha5.MachineNameLabelKey
```

Reply:
Interesting point of note here that @tzneal pointed out to me. Since the `lo.FromPtr` function is a generic function, it's an order of magnitude slower than the non-generic verison.

Reply:
We do a lot of filtering in our tight scheduling loop. Maybe worth specifying?